core/src/main/java/org/pentaho/di/cluster/SlaveConnectionManager.java

```diff
@@ -0,0 +1,155 @@
+/*! ******************************************************************************
+ *
+ * Pentaho Data Integration
+ *
+ * Copyright (C) 2002-2019 by Hitachi Vantara : http://www.pentaho.com
+ *
+ *******************************************************************************
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ ******************************************************************************/
+
+package org.pentaho.di.cluster;
+
+import org.apache.http.HttpHost;
+import org.apache.http.auth.AuthScope;
+import org.apache.http.auth.UsernamePasswordCredentials;
+import org.apache.http.client.CredentialsProvider;
+import org.apache.http.client.HttpClient;
+import org.apache.http.client.config.RequestConfig;
+import org.apache.http.impl.client.BasicCredentialsProvider;
+import org.apache.http.impl.client.HttpClientBuilder;
+import org.apache.http.impl.client.HttpClients;
+import org.apache.http.impl.conn.PoolingHttpClientConnectionManager;
+import org.pentaho.di.core.logging.LogChannel;
+
+import javax.net.ssl.KeyManager;
+import javax.net.ssl.SSLContext;
+import javax.net.ssl.X509TrustManager;
+import java.security.SecureRandom;
+import java.security.cert.CertificateException;
+import java.security.cert.X509Certificate;
+
+/**
+ * Encapsulates the Apache commons HTTP connection manager with a singleton. We can use this to limit the number of open
+ * connections to slave servers.
+ *
+ * @author matt
+ */
+public class SlaveConnectionManager {
+
+  private static final String SSL = "SSL";
+  private static final String KEYSTORE_SYSTEM_PROPERTY = "javax.net.ssl.keyStore";
+
+  private static SlaveConnectionManager slaveConnectionManager;
+
+  private PoolingHttpClientConnectionManager manager;
+
+  private SlaveConnectionManager() {
+    if ( needToInitializeSSLContext() ) {
+      try {
+        SSLContext context = SSLContext.getInstance( SSL );
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

An old version of SSL/TLS is enabled. We recommend that you enable TLS version 1.2, 1.3, or later. [Learn more](https://cheatsheetseries.owasp.org/cheatsheets/TLS_Cipher_String_Cheat_Sheet.html)